nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc

```diff
@@ -182,29 +209,17 @@ static int nntrainer_model_push_data(const GstTensorTrainerFramework *fw,
   }
 
   data->push_count++;
-  data->queue_count++;
   data->queue_rear = (data->queue_rear + 1) % data->queue_size;
-  ml_logd("front:%d, rear:%d, filled:%d", data->queue_front, data->queue_rear,
-          data->queue_count);
-
-  if (data->is_data_wait_locked && data->queue_count > 0) {
-    data->data_wait.notify_one();
-    ml_logd("send signal");
-  }
-
-  if (data->queue_count == data->queue_size) {
-    data->is_data_full_locked = TRUE;
-    ml_logd("locked, data is full");
-    std::unique_lock<std::mutex> lock(data->data_full_lock);
-    data->data_full.wait(lock);
-    ml_logd("unlocked, queue is empty");
-    data->is_data_full_locked = FALSE;
-  }
+  int queue_count = (data->queue_rear - data->queue_front + data->queue_size) %
+                    data->queue_size;
 
+  ml_logd("front:%d, rear:%d, filled:%d", data->queue_front, data->queue_rear,
+          queue_count);
   ml_logd("(pop/push: %d/%d)", data->pop_count, data->push_count);
   ml_logd("T-pushed: %d/%d, V-pushed:%d/%d\n",
           nntrainer->train_data->push_count, nntrainer->num_training_samples,
           nntrainer->valid_data->push_count, nntrainer->num_validation_samples);
+  data->data_empty.notify_one();
```

Comment:
Reference: https://en.cppreference.com/w/cpp/thread/condition_variable

```lock.unlock()``` should be done before notify_one

Reply:
The structure is as follows. 
Does this mean that queue_front and queue_rear require a mutex?
```
 std::mutex queue_lock;
 std::condition_variable data_full;
 std::condition_variable data_empty;

getSample() { // use queue data

  std::unique_lock<std::mutex> lock(queue_lock);
  data_empty.wait(lock, [this] { return !isQueueEmpty(); }); // Lock when the queue is empty
  [...]
  queue_front = (queue_front + 1) % queue_size;
  data_full.notify_one() // after pop data
}

push_data() { // write queue data
{
  std::unique_lock<std::mutex> lock(data->queue_lock);
  data->data_full.wait(lock, [&] { return !data->isQueueFull(); }); // Lock when the queue is full
 [...]
  data->queue_rear = (data->queue_rear + 1) % data->queue_size;
  data->data_empty.notify_one(); // after push data
}
```